arb_os/inbox.mini

```diff
@@ -157,19 +229,187 @@ impure func getFromL1Inbox(inbox: Inbox) -> Inbox {
+            timestamp: newMsgRaw.timestamp,
+            sender: newMsgRaw.sender,
+            requestId: newMsgRaw.requestId,
+            msgData: msgData,
+            provenance: struct {
+                l1SeqNum: newMsgRaw.requestId,
+                parentRequestId: 0,
+                indexInParent: ~0,
+            },
+        };
+    } else {
+        return inbox;
+    }
+
+    let blockNum = newMsg.blockNumber;
+    if (blockNum == 0) {
+        blockNum = seq.lastBlockNumSeen;
+    } else {
+        seq = seq with { lastBlockNumSeen: blockNum };
+    }
+    let timestamp = newMsg.timestamp;
+    if (timestamp == 0) {
+        timestamp = seq.lastBlockNumSeen;
+    } else {
+        seq = seq with { lastTimestampSeen: timestamp };
+    }
+
+    let isSequencerMsg = (newMsg.kind == const::L1MessageType_L2)
+                         && (bytearray_getByte(newMsg.msgData, 0) == const::L2MessageType_sequencerBatch)
+                         && (newMsg.sender == seq.sequencerAddr);
+    if (isSequencerMsg) {
+        // get low-order two bytes of blockNum assigned to this message, and reconstruct that blockNum
+        let bnum = (blockNum & ~0xffff)
+                         + asm (256-16, bytearray_get256(newMsg.msgData, 1)) uint { shr };
+        if ((bnum <= blockNum) || (bnum < 256*256)) {
+            blockNum = bnum;
+        } else {
+            blockNum = bnum-256*256;
+        }
+        // get low-order three bytes of timestamp assigned to this message, and reconstruct that timestamp
+        let ts = (timestamp & ~0xffffff)
+                        + asm (256-24, bytearray_get256(newMsg.msgData, 3)) uint { shr };
+        if ((ts <= timestamp) || (ts < 256*256*256)) {
+            timestamp = ts;
+        } else {
+            timestamp = ts - 256*256*256;
+        }
+
+        if (blockNum+seq.delayBlocks < seq.lastBlockNumSeen) {
+            blockNum = seq.lastBlockNumSeen - seq.delayBlocks;
+        }
+        if (timestamp+seq.delayTime < seq.lastTimestampSeen) {
+            timestamp = seq.lastTimestampSeen - seq.delayTime;
+        }
+
+        let (uSeq, uInboxQ) = releaseDelayedRequests(
+            seq,
+            inbox.queue,
+            blockNum,
+            timestamp,
+        );
+        seq = uSeq;
+        inbox = inbox with { queue: uInboxQ };
+
+        if (blockNum < seq.lastBlockNumReleased) {
+            blockNum = seq.lastBlockNumReleased;
+        }
+        if (timestamp < seq.lastTimestampReleased) {
+            timestamp = seq.lastTimestampReleased;
+        }
+
+        return inbox with {
+            queue: incomingRequestQueue_put(
+                inbox.queue,
+                newMsg with { blockNumber: blockNum } with { timestamp: timestamp }
+            )
+        } with {
+            sequencer: Some(
+                seq with {
+                    lastBlockNumReleased: blockNum
+                } with {
+                    lastTimestampReleased: timestamp
+                }
```

Comment:
Weren't these already adjusted by the call to `releaseDelayedRequests`?

Reply:
Not necessarily.  `releaseDelayedRequests` doesn't always set these fields.